utils/annotations.py

```diff
@@ -0,0 +1,421 @@
+from gnomad_hail import *
+
+
+def pop_max_expr(
+        freq: hl.expr.ArrayExpression,
+        freq_meta: hl.expr.ArrayExpression,
+        pops_to_exclude: Optional[Set[str]] = None
+) -> hl.expr.StructExpression:
+    """
+    Creates an expression containing popmax: the frequency information about the population
+    that has the highest AF from the populations provided in `freq_meta`,
+    excluding those specified in `pops_to_exclude`.
+    Only frequencies from adj populations are considered.
+    This resulting struct contains the following fields:
+    - AC: int32,
+    - AF: float64,
+    - AN: int32,
+    - homozygote_count: int32,
+    - pop: str
+
+    :param ArrayExpression freq: ArrayExpression of Structs with fields ['AC', 'AF', 'AN', 'homozygote_count']
+    :param ArrayExpression freq_meta: ArrayExpression of meta dictionaries corresponding to freq (as returned by annotate_freq)
+    :param set of str pops_to_exclude: Set of populations over which to calculate popmax
+    :return: Popmax struct
+    :rtype: StructExpression
+    """
+    _pops_to_exclude = hl.literal(pops_to_exclude)
+    popmax_freq_indices = hl.range(0, hl.len(freq_meta)).filter(
+        lambda i:
+        (hl.set(freq_meta[i].keys()) == {'group', 'pop'}) &
+        (freq_meta[i]['group'] == 'adj') &
+        (~_pops_to_exclude.contains(freq_meta[i]['pop']))
+    )
+    freq_filtered = popmax_freq_indices.map(
+        lambda i: freq[i].annotate(pop=freq_meta[i]['pop'])
+    ).filter(
+        lambda f: f.AC > 0
+    )
+
+    sorted_freqs = hl.sorted(freq_filtered, key=lambda x: x.AF, reverse=True)
+    return hl.or_missing(
+        hl.len(sorted_freqs) > 0,
+        sorted_freqs[0]
+    )
+
+
+def project_max_expr(
+        project_expr: hl.expr.StringExpression,
+        gt_expr: hl.expr.CallExpression,
+        alleles_expr: hl.expr.ArrayExpression,
+        n_projects: int = 5
+) -> hl.expr.ArrayExpression:
+    """
+    Creates an expression that computes allele frequency information by project for the `n_projects` with the largest AF at this row.
+    This return an array with one element per non-reference allele.
+    Each of these elements is itself an array of structs with the following fields:
+    - AC: int32,
+    - AF: float64,
+    - AN: int32,
+    - homozygote_count: int32,
+    - project: str
+
+    Notes
+    -----
+    Only projects with AF > 0 are returned.
+    In case of ties, the project ordering is no guaranteed, and at most `n_projects` are returned.
```

Comment:
correct, in case of a tie either project could be returned

Reply:
we could change this to returning all projects in this case, but  this could potentially lead to very large amounts of data returned. TBH, I don't think anyone uses this anymore...

Reply:
makes sense! is it worth keeping this function? (if so, maybe we should fix the small typo in `no guaranteed`)

Reply:
we should, and typo is fixed!